packages/nouns-contracts/contracts/NounsArt.sol

```diff
@@ -0,0 +1,449 @@
+        return Inflate.puff(source, destlen);
+    }
+
+    function _addBackground(string calldata _background) internal {
+        backgrounds.push(_background);
+    }
+
+    function addPage(
+        Trait storage trait,
+        bytes calldata encodedCompressed,
+        uint80 decompressedLength,
+        uint16 imageCount
+    ) internal {
+        if (encodedCompressed.length == 0) {
+            revert EmptyBytes();
+        }
+        address pointer = SSTORE2.write(encodedCompressed);
+        addPage(trait, pointer, decompressedLength, imageCount);
+    }
+
+    function addPage(
+        Trait storage trait,
+        address pointer,
+        uint80 decompressedLength,
+        uint16 imageCount
+    ) internal {
+        if (decompressedLength == 0) {
+            revert BadDecompressedLength();
+        }
+        if (imageCount == 0) {
+            revert BadImageCount();
+        }
+        trait.storagePages.push(
+            NounArtStoragePage({ pointer: pointer, decompressedLength: decompressedLength, imageCount: imageCount })
+        );
+        trait.storedImagesCount += imageCount;
+    }
+
+    function imageByIndex(INounsArt.Trait storage trait, uint256 index) internal view returns (bytes memory) {
+        (INounsArt.NounArtStoragePage storage page, uint256 indexInPage) = getPage(trait.storagePages, index);
+        bytes[] memory decompressedImages = decompressAndDecode(page);
+        return decompressedImages[indexInPage];
+    }
+
+    /**
+     * @dev Given an image index, this function finds the storage page the image is in, and the relative index
+     * inside the page, so the image can be read from storage.
+     * Example: if you have 2 pages with 100 images each, and you want to get image 150, this function would return
+     * the 2nd page, and the 50th index.
+     * @return INounsArt.NounArtStoragePage the page containing the image at index
+     * @return uint256 the index of the image in the page
+     */
+    function getPage(INounsArt.NounArtStoragePage[] storage pages, uint256 index)
+        internal
+        view
+        returns (INounsArt.NounArtStoragePage storage, uint256)
+    {
+        uint256 len = pages.length;
+        uint256 pageFirstImageIndex = 0;
+        for (uint256 i = 0; i < len; i++) {
+            INounsArt.NounArtStoragePage storage page = pages[i];
+
+            if (index < pageFirstImageIndex + page.imageCount) {
+                return (page, index - pageFirstImageIndex);
+            }
+
+            pageFirstImageIndex += page.imageCount;
+        }
+
+        revert ImageNotFound();
+    }
+
+    function decompressAndDecode(INounsArt.NounArtStoragePage storage page) internal view returns (bytes[] memory) {
+        bytes memory compressedData = SSTORE2.read(page.pointer);
+        (, bytes memory decompressedData) = Inflate.puff(compressedData, page.decompressedLength);
```

Comment:
What was the added gas overhead when converting `Inflate.puff` to an external call? I've forgotten where it was posted. Thinking over it again, would be nice to gas-golf the inflate implementation over time.

Reply:
I pushed another test (here)[https://github.com/eladmallel/nouns-descriptor-benchmark] and the overhead seems to be 600K.

Reply:
thanks for the great question! given that `tokenURI` is in the ~100M gas ballpark, I think adding 600K gas should be ok. So I am leaning towards deploying Inflate as a contract dependency we can upgrade over time.

@solimander @davidbrai wdyt?